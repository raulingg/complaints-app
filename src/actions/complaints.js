import database from '../firebase/firebase';

export const addComplaint = complaint => ({
  type: 'ADD_COMPLAINT',
  complaint,
});

export const startAddComplaint = (complaintData = {}) => (dispatch, getState) => {
  const {
    auth: { uid },
  } = getState();
  const {
    title = '',
    content = '',
    reportedAt = 0,
    happenedAt = 0,
    reportTo = '',
    address = '',
  } = complaintData;
  const complaint = { title, content, reportedAt, happenedAt, reportTo, address, uid };

  return database
    .ref('complaints')
    .push(complaint)
    .then(ref => {
      dispatch(
        addComplaint({
          id: ref.key,
          ...complaint,
        })
      );
    });
};

export const editComplaint = (id, updates) => ({
  type: 'EDIT_COMPLAINT',
  id,
  updates,
});

export const startEditComplaint = (id, updates) => dispatch =>
  database
    .ref(`complaints/${id}`)
    .update(updates)
    .then(() => {
      dispatch(editComplaint(id, updates));
    });

export const setComplaints = complaints => ({
  type: 'SET_COMPLAINTS',
  complaints,
});

export const startSetComplaints = () => dispatch =>
  database
    .ref('complaints')
    .once('value')
    .then(snapshot => {
      const complaints = [];

      snapshot.forEach(childSnapshot => {
        complaints.push({
          id: childSnapshot.key,
          ...childSnapshot.val(),
        });
      });

      dispatch(setComplaints(complaints));
    });
