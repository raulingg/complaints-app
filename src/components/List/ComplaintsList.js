import React from 'react';
import { connect } from 'react-redux';

export const ComplaintsList = props => {
  const { complaints } = props;

  return (
    <div>
      <div>
        <div>Denuncia</div>
        <div>Tipo</div>
        <div>Fecha</div>
      </div>
      <div>
        {complaints.length === 0 ? (
          <div>
            <span>No encontramos ninguna denuncia.</span>
          </div>
        ) : (
          complaints.map(complaint => (
            <div key={complaint.id}>
              <p>
                <div>{complaint.title}</div>
                <div>{complaint.content}</div>
              </p>
            </div>
          ))
        )}
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  complaints: state.complaints,
});

export default connect(mapStateToProps)(ComplaintsList);
