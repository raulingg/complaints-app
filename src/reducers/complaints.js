// Expenses Reducer

const complaintsReducerDefaultState = [];

export default (state = complaintsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_COMPLAINT':
      return [...state, action.complaint];
    case 'REMOVE_COMPLAINT':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_COMPLAINT':
      return state.map(complaint => {
        if (complaint.id === action.id) {
          return {
            ...complaint,
            ...action.updates,
          };
        }
        return complaint;
      });
    case 'SET_COMPLAINTS':
      return action.complaints;
    default:
      return state;
  }
};
