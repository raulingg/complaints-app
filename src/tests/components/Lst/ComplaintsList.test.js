import React from 'react';
import { shallow } from 'enzyme';
import { ComplaintsList } from '../../../components/List/ComplaintsList';
import complaints from '../../fixtures/complaints';

test('should correctly render ComplaintsList', () => {
  const wrapper = shallow(<ComplaintsList complaints={complaints} />);
  expect(wrapper).toMatchSnapshot();
});
