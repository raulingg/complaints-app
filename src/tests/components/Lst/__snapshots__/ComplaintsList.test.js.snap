// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ComplaintsList 1`] = `
<div>
  <div>
    <div>
      Denuncia
    </div>
    <div>
      Tipo
    </div>
    <div>
      Fecha
    </div>
  </div>
  <div>
    <div>
      <div>
        Me robaron mi celular
      </div>
      <div>
        esta yo en la calle cuando de repente un maricon me robo el celular.
      </div>
    </div>
    <div>
      <div>
        Me robaron mi celular
      </div>
      <div>
        esta yo en la calle cuando de repente un maricon me robo el celular.
      </div>
    </div>
    <div>
      <div>
        Me robaron mi celular
      </div>
      <div>
        esta yo en la calle cuando de repente un maricon me robo el celular.
      </div>
    </div>
  </div>
</div>
`;
